FPU exceptions (QEMU test)

Periodically, as you guys are working on the emulation code, I retest the known failures in the test suite to see if anything got magically fixed by the changes you're making. Of course the answer is usually no, if I do find any I'll unmark the 'known failure' so the test gets run from then on and we'll maintain your progress.
Since you've been looking at FPU exceptions, perhaps you'd like to have a glance at this one?
~~~
$ env NO_FAILFAST=1 NO_ACTIONS=1 ./test/test_processor.py KVMTestCase.test_fpu_fp_exceptions_kvm_jit
 
Using KVM
Test MS-DOS-6.22 FPU fp-exceptions KVM vm86 + JIT DPMI                           ... FAIL

======================================================================
FAIL: Test MS-DOS-6.22 FPU fp-exceptions KVM vm86 + JIT DPMI                          
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/clients/common/dosemu2.git/test/fpu/qemu.py", line 93, in do_test
    _dotest(self, *test)
  File "/clients/common/dosemu2.git/test/fpu/qemu.py", line 88, in _dotest
    self.assertIn("PASS:", results)
AssertionError: 'PASS:' not found in 'testit.bat

C:\>c:\fputest
Exiting due to signal SIGILL 
Invalid Opcode at eip=00002f86
eax=00000001 ebx=00094980 ecx=00094760 edx=fff80000 esi=00094963 edi=0000000c
ebp=00094748 esp=00094710 program=c:\FPUTEST.EXE
cs: sel=010f  base=08034000  limit=0009ffff
ds: sel=0117  base=08034000  limit=0009ffff
es: sel=0117  base=08034000  limit=0009ffff
fs: sel=00b7  base=00005200  limit=0000ffff
gs: sel=0127  base=00000000  limit=0010ffff
ss: sel=0117  base=08034000  limit=0009ffff
App stack: [000947d0..000147d0]  Exceptn stack: [0001477c..0001283c]

Call frame traceback EIPs:
  0x00002f86
  0x0000729d

C:\>'
Further info in file '/clients/common/dosemu2.git/test_processor.KVMTestCase.test_fpu_fp_exceptions_kvm_jit.log'
Further info in file '/clients/common/dosemu2.git/test_processor.KVMTestCase.test_fpu_fp_exceptions_kvm_jit.xpt'

----------------------------------------------------------------------
Ran 1 test in 1.006s

FAILED (failures=1)
~~~

The SIM version of it `test_fpu_fp_exceptions_kvm_sim` also fails, but there seems to other things wrong with that before an 'Invalid Opcode' brings it to a halt.

--- a/test/fpu/test-i386-fp-exceptions.c
+++ b/test/fpu/test-i386-fp-exceptions.c
@@ -537,133 +537,135 @@ int main(void)
         ret = 1;
     }
 
-    __asm__ volatile ("fnclex");
-    __asm__ volatile ("fisttps %0" : "=m" (res_16) : "t" (1.5L) : "st");
-    __asm__ volatile ("fnstsw" : "=a" (sw));
-    if ((sw & EXC) != PE) {
-        printf("FAIL: fisttp inexact\n");
-        ret = 1;
-    }
-    __asm__ volatile ("fnclex");
-    __asm__ volatile ("fisttps %0" : "=m" (res_16) : "t" (32768.0L) : "st");
-    __asm__ volatile ("fnstsw" : "=a" (sw));
-    if ((sw & EXC) != IE) {
-        printf("FAIL: fisttp 32768\n");
-        ret = 1;
-    }
-    __asm__ volatile ("fnclex");
-    __asm__ volatile ("fisttps %0" : "=m" (res_16) : "t" (32768.5L) : "st");
-    __asm__ volatile ("fnstsw" : "=a" (sw));
-    if ((sw & EXC) != IE) {
-        printf("FAIL: fisttp 32768.5\n");
-        ret = 1;
-    }
-    __asm__ volatile ("fnclex");
-    __asm__ volatile ("fisttps %0" : "=m" (res_16) : "t" (-32769.0L) : "st");
-    __asm__ volatile ("fnstsw" : "=a" (sw));
-    if ((sw & EXC) != IE) {
-        printf("FAIL: fisttp -32769\n");
-        ret = 1;
-    }
-    __asm__ volatile ("fnclex");
-    __asm__ volatile ("fisttps %0" : "=m" (res_16) : "t" (-32769.5L) : "st");
-    __asm__ volatile ("fnstsw" : "=a" (sw));
-    if ((sw & EXC) != IE) {
-        printf("FAIL: fisttp -32769.5\n");
-        ret = 1;
-    }
-    __asm__ volatile ("fnclex");
-    __asm__ volatile ("fisttps %0" : "=m" (res_16) : "t" (ld_nan) : "st");
-    __asm__ volatile ("fnstsw" : "=a" (sw));
-    if ((sw & EXC) != IE) {
-        printf("FAIL: fisttp nan\n");
-        ret = 1;
-    }
-    __asm__ volatile ("fnclex");
-    __asm__ volatile ("fisttps %0" : "=m" (res_16) : "t" (ld_invalid_1.ld) :
-                      "st");
-    __asm__ volatile ("fnstsw" : "=a" (sw));
-    if ((sw & EXC) != IE) {
-        printf("FAIL: fisttp invalid\n");
-        ret = 1;
-    }
+    if (__builtin_cpu_supports("sse3")) {
+        __asm__ volatile ("fnclex");
+        __asm__ volatile ("fisttps %0" : "=m" (res_16) : "t" (1.5L) : "st");
+        __asm__ volatile ("fnstsw" : "=a" (sw));
+        if ((sw & EXC) != PE) {
+            printf("FAIL: fisttp inexact\n");
+            ret = 1;
+        }
+        __asm__ volatile ("fnclex");
+        __asm__ volatile ("fisttps %0" : "=m" (res_16) : "t" (32768.0L) : "st");
+        __asm__ volatile ("fnstsw" : "=a" (sw));
+        if ((sw & EXC) != IE) {
+            printf("FAIL: fisttp 32768\n");
+            ret = 1;
+        }
+        __asm__ volatile ("fnclex");
+        __asm__ volatile ("fisttps %0" : "=m" (res_16) : "t" (32768.5L) : "st");
+        __asm__ volatile ("fnstsw" : "=a" (sw));
+        if ((sw & EXC) != IE) {
+            printf("FAIL: fisttp 32768.5\n");
+            ret = 1;
+        }
+        __asm__ volatile ("fnclex");
+        __asm__ volatile ("fisttps %0" : "=m" (res_16) : "t" (-32769.0L) : "st");
+        __asm__ volatile ("fnstsw" : "=a" (sw));
+        if ((sw & EXC) != IE) {
+            printf("FAIL: fisttp -32769\n");
+            ret = 1;
+        }
+        __asm__ volatile ("fnclex");
+        __asm__ volatile ("fisttps %0" : "=m" (res_16) : "t" (-32769.5L) : "st");
+        __asm__ volatile ("fnstsw" : "=a" (sw));
+        if ((sw & EXC) != IE) {
+            printf("FAIL: fisttp -32769.5\n");
+            ret = 1;
+        }
+        __asm__ volatile ("fnclex");
+        __asm__ volatile ("fisttps %0" : "=m" (res_16) : "t" (ld_nan) : "st");
+        __asm__ volatile ("fnstsw" : "=a" (sw));
+        if ((sw & EXC) != IE) {
+            printf("FAIL: fisttp nan\n");
+            ret = 1;
+        }
+        __asm__ volatile ("fnclex");
+        __asm__ volatile ("fisttps %0" : "=m" (res_16) : "t" (ld_invalid_1.ld) :
+                          "st");
+        __asm__ volatile ("fnstsw" : "=a" (sw));
+        if ((sw & EXC) != IE) {
+            printf("FAIL: fisttp invalid\n");
+            ret = 1;
+        }
 
-    __asm__ volatile ("fnclex");
-    __asm__ volatile ("fisttpl %0" : "=m" (res_32) : "t" (1.5L) : "st");
-    __asm__ volatile ("fnstsw" : "=a" (sw));
-    if ((sw & EXC) != PE) {
-        printf("FAIL: fisttpl inexact\n");
-        ret = 1;
-    }
-    __asm__ volatile ("fnclex");
-    __asm__ volatile ("fisttpl %0" : "=m" (res_32) : "t" (2147483648.0L) :
-                      "st");
-    __asm__ volatile ("fnstsw" : "=a" (sw));
-    if ((sw & EXC) != IE) {
-        printf("FAIL: fisttpl 2147483648\n");
-        ret = 1;
-    }
-    __asm__ volatile ("fnclex");
-    __asm__ volatile ("fisttpl %0" : "=m" (res_32) : "t" (-2147483649.0L) :
-                      "st");
-    __asm__ volatile ("fnstsw" : "=a" (sw));
-    if ((sw & EXC) != IE) {
-        printf("FAIL: fisttpl -2147483649\n");
-        ret = 1;
-    }
-    __asm__ volatile ("fnclex");
-    __asm__ volatile ("fisttpl %0" : "=m" (res_32) : "t" (ld_nan) : "st");
-    __asm__ volatile ("fnstsw" : "=a" (sw));
-    if ((sw & EXC) != IE) {
-        printf("FAIL: fisttpl nan\n");
-        ret = 1;
-    }
-    __asm__ volatile ("fnclex");
-    __asm__ volatile ("fisttpl %0" : "=m" (res_32) : "t" (ld_invalid_1.ld) :
-                      "st");
-    __asm__ volatile ("fnstsw" : "=a" (sw));
-    if ((sw & EXC) != IE) {
-        printf("FAIL: fisttpl invalid\n");
-        ret = 1;
-    }
+        __asm__ volatile ("fnclex");
+        __asm__ volatile ("fisttpl %0" : "=m" (res_32) : "t" (1.5L) : "st");
+        __asm__ volatile ("fnstsw" : "=a" (sw));
+        if ((sw & EXC) != PE) {
+            printf("FAIL: fisttpl inexact\n");
+            ret = 1;
+        }
+        __asm__ volatile ("fnclex");
+        __asm__ volatile ("fisttpl %0" : "=m" (res_32) : "t" (2147483648.0L) :
+                          "st");
+        __asm__ volatile ("fnstsw" : "=a" (sw));
+        if ((sw & EXC) != IE) {
+            printf("FAIL: fisttpl 2147483648\n");
+            ret = 1;
+        }
+        __asm__ volatile ("fnclex");
+        __asm__ volatile ("fisttpl %0" : "=m" (res_32) : "t" (-2147483649.0L) :
+                          "st");
+        __asm__ volatile ("fnstsw" : "=a" (sw));
+        if ((sw & EXC) != IE) {
+            printf("FAIL: fisttpl -2147483649\n");
+            ret = 1;
+        }
+        __asm__ volatile ("fnclex");
+        __asm__ volatile ("fisttpl %0" : "=m" (res_32) : "t" (ld_nan) : "st");
+        __asm__ volatile ("fnstsw" : "=a" (sw));
+        if ((sw & EXC) != IE) {
+            printf("FAIL: fisttpl nan\n");
+            ret = 1;
+        }
+        __asm__ volatile ("fnclex");
+        __asm__ volatile ("fisttpl %0" : "=m" (res_32) : "t" (ld_invalid_1.ld) :
+                          "st");
+        __asm__ volatile ("fnstsw" : "=a" (sw));
+        if ((sw & EXC) != IE) {
+            printf("FAIL: fisttpl invalid\n");
+            ret = 1;
+        }
 
-    __asm__ volatile ("fnclex");
-    __asm__ volatile ("fisttpll %0" : "=m" (res_64) : "t" (1.5L) : "st");
-    __asm__ volatile ("fnstsw" : "=a" (sw));
-    if ((sw & EXC) != PE) {
-        printf("FAIL: fisttpll inexact\n");
-        ret = 1;
-    }
-    __asm__ volatile ("fnclex");
-    __asm__ volatile ("fisttpll %0" : "=m" (res_64) : "t" (0x1p63) :
-                      "st");
-    __asm__ volatile ("fnstsw" : "=a" (sw));
-    if ((sw & EXC) != IE) {
-        printf("FAIL: fisttpll 0x1p63\n");
-        ret = 1;
-    }
-    __asm__ volatile ("fnclex");
-    __asm__ volatile ("fisttpll %0" : "=m" (res_64) : "t" (-0x1.1p63L) :
-                      "st");
-    __asm__ volatile ("fnstsw" : "=a" (sw));
-    if ((sw & EXC) != IE) {
-        printf("FAIL: fisttpll -0x1.1p63\n");
-        ret = 1;
-    }
-    __asm__ volatile ("fnclex");
-    __asm__ volatile ("fisttpll %0" : "=m" (res_64) : "t" (ld_nan) : "st");
-    __asm__ volatile ("fnstsw" : "=a" (sw));
-    if ((sw & EXC) != IE) {
-        printf("FAIL: fisttpll nan\n");
-        ret = 1;
-    }
-    __asm__ volatile ("fnclex");
-    __asm__ volatile ("fisttpll %0" : "=m" (res_64) : "t" (ld_invalid_1.ld) :
-                      "st");
-    __asm__ volatile ("fnstsw" : "=a" (sw));
-    if ((sw & EXC) != IE) {
-        printf("FAIL: fisttpll invalid\n");
-        ret = 1;
+        __asm__ volatile ("fnclex");
+        __asm__ volatile ("fisttpll %0" : "=m" (res_64) : "t" (1.5L) : "st");
+        __asm__ volatile ("fnstsw" : "=a" (sw));
+        if ((sw & EXC) != PE) {
+            printf("FAIL: fisttpll inexact\n");
+            ret = 1;
+        }
+        __asm__ volatile ("fnclex");
+        __asm__ volatile ("fisttpll %0" : "=m" (res_64) : "t" (0x1p63) :
+                          "st");
+        __asm__ volatile ("fnstsw" : "=a" (sw));
+        if ((sw & EXC) != IE) {
+            printf("FAIL: fisttpll 0x1p63\n");
+            ret = 1;
+        }
+        __asm__ volatile ("fnclex");
+        __asm__ volatile ("fisttpll %0" : "=m" (res_64) : "t" (-0x1.1p63L) :
+                          "st");
+        __asm__ volatile ("fnstsw" : "=a" (sw));
+        if ((sw & EXC) != IE) {
+            printf("FAIL: fisttpll -0x1.1p63\n");
+            ret = 1;
+        }
+        __asm__ volatile ("fnclex");
+        __asm__ volatile ("fisttpll %0" : "=m" (res_64) : "t" (ld_nan) : "st");
+        __asm__ volatile ("fnstsw" : "=a" (sw));
+        if ((sw & EXC) != IE) {
+            printf("FAIL: fisttpll nan\n");
+            ret = 1;
+        }
+        __asm__ volatile ("fnclex");
+        __asm__ volatile ("fisttpll %0" : "=m" (res_64) : "t" (ld_invalid_1.ld) :
+                          "st");
+        __asm__ volatile ("fnstsw" : "=a" (sw));
+        if ((sw & EXC) != IE) {
+            printf("FAIL: fisttpll invalid\n");
+            ret = 1;
+        }
     }
 
     __asm__ volatile ("fnclex");

--- a/test/test_processor.py
+++ b/test/test_processor.py
@@ -2,7 +2,8 @@
 
 from subprocess import check_call
 
-from common_framework import BaseTestCase, main, main_setup, mark, KNOWNFAIL
+from common_framework import (BaseTestCase, main, main_setup, mark,
+                              KNOWNFAIL, UNSUPPORTED)
 from common_os import ppdosgit
 
 from func_build_freecom import build_freecom
@@ -38,9 +39,8 @@ class OurTestCase(BaseTestCase):
 
 class KVMTestCase(ppdosgit(OurTestCase, {
         "test_fpu_f2xm1_kvm_sim": KNOWNFAIL,
-        "test_fpu_fisttp_kvm_jit": KNOWNFAIL,
-        "test_fpu_fisttp_kvm_sim": KNOWNFAIL,
-        "test_fpu_fp_exceptions_kvm_jit": KNOWNFAIL,
+        "test_fpu_fisttp_kvm_jit": UNSUPPORTED,  # Requires Pentium 4 (SSE3)
+        "test_fpu_fisttp_kvm_sim": UNSUPPORTED,  # Requires Pentium 4 (SSE3)
         "test_fpu_fp_exceptions_kvm_sim": KNOWNFAIL,
         "test_fpu_fyl2x_kvm_sim": KNOWNFAIL,
         "test_fpu_fyl2xp1_kvm_sim": KNOWNFAIL,
@@ -51,9 +51,8 @@ class KVMTestCase(ppdosgit(OurTestCase, {
 
 class EMUTestCase(ppdosgit(OurTestCase, {
         "test_fpu_f2xm1_sim_sim": KNOWNFAIL,
-        "test_fpu_fisttp_jit_jit": KNOWNFAIL,
-        "test_fpu_fisttp_sim_sim": KNOWNFAIL,
-        "test_fpu_fp_exceptions_jit_jit": KNOWNFAIL,
+        "test_fpu_fisttp_jit_jit": UNSUPPORTED,  # Requires Pentium 4 (SSE3)
+        "test_fpu_fisttp_sim_sim": UNSUPPORTED,  # Requires Pentium 4 (SSE3)
         "test_fpu_fp_exceptions_sim_sim": KNOWNFAIL,
         "test_fpu_fyl2x_sim_sim": KNOWNFAIL,
         "test_fpu_fyl2xp1_sim_sim": KNOWNFAIL,

